README.md

# TeeStreaming

This .NET assembly implements a TeeStream, which is a stream construct which, when written to, can output the data to multiple sink streams.
The TeeStream can also be used as a readable stream itself.

## Usage

To create a TeeStream, it's consructor with the sink streams you want to redirect output to.

```c#
async Task RedirectStream(Stream source)
{
    using (var dest1 = new FileStream("/foo"))
    using (var dest2 = new MemoryStream())
    using (var tee = new TeeStram(dest1, dest2))
    {
        await source.CopyToAsync(tee);
        // ...
    }
}
```

You can also specify a buffer size to facilitate larger read/write blocks than the default of 81,920.

```c#
async Task RedirectStream(Stream source)
{
    using (var dest1 = new FileStream("/foo"))
    using (var dest2 = new MemoryStream())
    using (var tee = new TeeStram(256000, dest1, dest2))
    {
        await source.CopyToAsync(tee);
        // ...
    }
}
```

> **Always be sure to Close or Dispose of this tream when done with it to automatically flush any buffered contents.**


### Working With Non-Seekable Streams - **SetAtEnd()**

The **SetAtEnd()** method indicates that the source stream has been completely
read.  The default **CopyTo**/**CopyToAsync** methods simply read from the 
source and write to the destination stream in blocks.  However, there is no way for the destination stream to "know" that the source has been completely read.

This matters for TeeStream because it needs to return zero from **Read**/**ReadAsync** when sink streams are reading from it, once the source stream
has been consumed.

The **SetAtEnd** method instructs TeeStream to return zero on 
Read/ReadAsync, which lets the sinks know that the source stream is done.

```c#
async Task RedirectStream(Stream source)
{
    using (var dest1 = new FileStream("/foo"))
    using (var dest2 = new MemoryStream())
    using (var tee = new TeeStram(256000, dest1, dest2))
    {
        await source.CopyToAsync(tee);
        tee.SetAtEnd();
    }
}
```

### Optimized Copying - **CopyFromAsync()**

TeeStream's **CopyFromAsync()** method is the reverse of **CopyToAsync**, in that the method will pull all data from the source stream and know when it is completely done.  

```c#
async Task RedirectStream(Stream source)
{
    using (var dest1 = new FileStream("/foo"))
    using (var dest2 = new MemoryStream())
    using (var tee = new TeeStram(256000, dest1, dest2))
    {
        await tee.CopyFromAsync(source);
    }
}
```

This method deploys a couple of tricks to enhance performance.  First, it will buffer incoming data and then write it out in bigger chunks to the sinks.  If you are pulling data from a stream that can only be read in small chunks (like output from a console application), this helps by consolidating write calls to the sinks.  Also, reads and writes (when ready) are triggered simultaneously as opposed to sequentially.

### Using the TeeStream as an Input

You can use the TeeStream as an input by specifing **TeeStream.Self** as a sink.  See the following example.

## Example: Thumbnail Generation

Assume you are working on an application where somebody uploads an image file via HTTP post in ASP.NET, with the Response.Body a stream to that file data.  You want to save that image to storage, generate a thumbnail, save the thumbail to storage, and return the thumbnail as the Response.Body.  

Furthermore, because the image may be a PDF, you want to use ImageMagick's **convert** utility.

You can do this without temp files or large MemoryStreams by doing the following:

```c#
async Task SaveImage(Stream requestBody, Stream responseBody, string imageFileName, string thumbnailFileName)
{
    using (var fsImage = new FileStream(imageFileName, FileMode.Create))
    using (var fsThumbnail = new FileStream(thumbnailFileName, FileMode.Create))
    using (var teeThumbnail = new TeeStream(fsThumbnail, reesponseBody))
    using (var teeInput = new FileStream(fsImage, TeeStream.Self))
    {
        var taskGenerate = Task.Run(async () => {
            var info = new ProcessStartInfo
            {
                FileName = "convert",
                Arguments = $"-resize \"200x200\" - jpg:-",
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };
            var process = Process.Start(info);
            if (process == null) {
                throw new Exception($"Unable to start \"convert");
            }

            await teeInput.CopyToAsync(process.StandardInput.BaseStream);
            process.StandardInput.Close();
            await process.StandardOutput.BaseStream.CopyToAsync(teeThumbnail);
            var stderr = await process.StandardError.ReadToEndAsync();
            await process.WaitForExitAsync();

            if (process.ExitCode != 0) {
                throw new Exception(stderr.Length > 0 ? stderr : $"Error converting image, exit code: {process.ExitCode}");
            }        
        });
        await teeInput.CopyFromAsync(requestBody);
        taskGenerate.Wait();
    }
}
```

What's happening?

1. We are creating file streams to save the image and thumbnail to.  You can write to any stream.
2. We create **teeInput** stream, which will output to a file for the full sized image and also allow the tee stream itself to be read from.  This will be fed into ImageMagick's convert utility.
3. We create a **teeThumbnail** stream that will direct data to a thumbnail file and the response body.
4. We create a task to call the ImageMagick convert utility, set to read from STDIN and write to STDOUT.  We will redirect both STDIN and STDOUT.  We will copy **teeInput** to STDIN, and copy STDOUT to **teeThumbnail**.  Importantly, we launch this as a Task but do not wait for it, yet.
5. We then copy the request body to **teeInput**.  The task we created above will read this data from **teeInput** and, via **teeThumbnail**, save the thumbnail to a file and the response body stream
6.  Finally, we wait for the task to complete.

## Demonstration

There is a project on [GitHub](https://github.com/jasonterando/TeeStreamingDemo) demonstrating how to tee stream uploads to S3.

CHANGELOG.md
# Changelog

## 0.1.0
Initial commit

## 0.1.1
Remove "implicit using" setting

## 0.1.2
Update project URL in csproj
Remove test/bin and test/obj files

## 0.1.3
Add reference to demonstration project

test/runtests.sh
#!/bin/sh
mkdir -p .results
rm -r -f .results/*
dotnet test --collect:"XPlat Code Coverage" -r .results/log -l trx
result=$?
dotnet reportgenerator -reports:`find .results -name coverage.cobertura.xml` -targetdir:.results/reports
if [ $result = 0 ]; then
    echo "All tests passed!"
else
    echo "Some tests failed :("
fi
exit $result

test/TeeStreamTests.cs
using System.Buffers;
using NUnit.Framework;
using TeeStreaming;
using System.Threading;
using System.Threading.Tasks;
using System.IO;
using System;
using Moq;

namespace test;

class TeeStreamTests
{
    [Test]
    public void CanCreateWithDefaults()
    {
        using (var testStream = new MemoryStream(1))
        {
            using (var tee = new TeeStream(testStream))
            {
                Assert.AreEqual(81920, tee._bufferSize);
                Assert.Null(tee._outputBuffer);
                Assert.AreEqual(new Stream[] { testStream }, tee._sinks);
            }
        }
    }

    [Test]
    public void CanCreateWithSelf()
    {
        using (var testStream = new MemoryStream(1))
        {
            using (var tee = new TeeStream(TeeStream.Self, testStream))
            {
                Assert.AreEqual(81920, tee._bufferSize);
                Assert.NotNull(tee._outputBuffer);
                Assert.AreEqual(new Stream[] { testStream }, tee._sinks);
            }
        }
    }

    [Test]
    public void CreateFaultsOnNoSinks()
    {
        Assert.Throws<ArgumentException>(() => new TeeStream(new Stream[] { }),
            "There must be at least one sink");
    }

    [Test]
    public void CreateFaultsOnSmallBuffer()
    {
        Assert.Throws<ArgumentException>(() => new TeeStream(128, TeeStream.Self),
            "Buffer size must be at least 1,024");
    }

    [Test]
    public void CreateFaultsOnWritableSink()
    {
        var sink = new Mock<Stream>();
        sink.SetupGet(s => s.CanWrite).Returns(false);
        Assert.Throws<ArgumentException>(() => new TeeStream(sink.Object),
            "Sink #1 is not writable");
    }

    [Test]
    public void CloseCallsSinkClose()
    {
        var sink = new Mock<Stream>();
        sink.SetupGet(s => s.CanWrite).Returns(true);
        sink.Setup(s => s.Close()).Verifiable();
        using (var tee = new TeeStream(sink.Object))
        {
            tee.Close();
            sink.Verify(s => s.Close(), Times.Once);
        }
    }

    [Test]
    public void DisposeDisposeesOfOutputBuffer()
    {
        var tee = new TeeStream(TeeStream.Self);
        Assert.NotNull(tee._outputBuffer);
        tee.Dispose();
        Assert.Null(tee._outputBuffer);
    }

    [Test]
    public async Task CopyFromAsyncReadsAndWritesEntireStreamWithSmallBuffer()
    {
        var sampleBuffer = new byte[4096];
        var rnd = new Random();
        rnd.NextBytes(sampleBuffer);

        using (var source = new MemoryStream(sampleBuffer))
        using (var destination = new MemoryStream(sampleBuffer.Length))
        using (var tee = new TeeStream(sampleBuffer.Length / 4, destination))
        {
            source.Position = 0;
            await tee.CopyFromAsync(source);

            destination.Position = 0;
            var results = new byte[sampleBuffer.Length];
            await destination.ReadAsync(results, 0, sampleBuffer.Length);
            Assert.AreEqual(sampleBuffer, results);
            Assert.AreEqual(sampleBuffer.Length, destination.Length);
        }
    }

    [Test]
    public async Task CopyFromAsyncReadsAndWritesEntireStreamWithLargeBuffer()
    {
        var sampleBuffer = new byte[1024];
        var rnd = new Random();
        rnd.NextBytes(sampleBuffer);

        using (var source = new MemoryStream(sampleBuffer))
        using (var destination = new MemoryStream(sampleBuffer.Length))
        using (var tee = new TeeStream(sampleBuffer.Length * 2, destination))
        {
            source.Position = 0;
            await tee.CopyFromAsync(source);

            destination.Position = 0;
            var results = new byte[sampleBuffer.Length];
            await destination.ReadAsync(results, 0, sampleBuffer.Length);
            Assert.AreEqual(sampleBuffer, results);
            Assert.AreEqual(sampleBuffer.Length, destination.Length);
        }
    }

    [Test]
    public async Task CopyFromAsyncReadsAndWritesEntireStreamWithSmallData()
    {
        var sampleBuffer = new byte[128];
        var rnd = new Random();
        rnd.NextBytes(sampleBuffer);

        using (var source = new MemoryStream(sampleBuffer))
        using (var destination = new MemoryStream(sampleBuffer.Length))
        using (var tee = new TeeStream(destination))
        {
            source.Position = 0;
            await tee.CopyFromAsync(source);

            destination.Position = 0;
            var results = new byte[sampleBuffer.Length];
            await destination.ReadAsync(results, 0, sampleBuffer.Length);
            Assert.AreEqual(sampleBuffer, results);
            Assert.AreEqual(sampleBuffer.Length, destination.Length);
        }
    }    

    [Test]
    public void CopyFromAsyncCancels()
    {
        var sampleBuffer = new byte[1024];
        var rnd = new Random();
        rnd.NextBytes(sampleBuffer);

        var cancel = new CancellationTokenSource();

        using (var source = new MemoryStream(sampleBuffer))
        using (var destination = new MemoryStream(sampleBuffer.Length))
        using (var tee = new TeeStream(sampleBuffer.Length * 2, destination))
        {
            source.Position = 0;
            cancel.Cancel();
            Assert.CatchAsync<AggregateException>(async () => await tee.CopyFromAsync(source, cancel.Token));
        }
    }

    [Test]
    public void SetAtEndSetsValue()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            tee.SetAtEnd();
            Assert.AreEqual(true, tee._atEnd);
        }
    }

    [Test]
    public void CanReadTrueIfSelfBuffering()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.AreEqual(true, tee.CanRead);
        }
    }

    [Test]
    public void CanReadFalseIfNotSelfBuffering()
    {
        using (var source = new MemoryStream(1))
        using (var tee = new TeeStream(source))
        {
            Assert.AreEqual(false, tee.CanRead);
        }
    }

    [Test]
    public void ReadAsyncThrowsIfNotSelfBuffering()
    {
        using (var source = new MemoryStream(1))
        using (var tee = new TeeStream(source))
        {
            var buffer = new byte[1];
            Assert.ThrowsAsync<NotImplementedException>(async () => await tee.ReadAsync(buffer, 0, 1));
        }
    }

    [Test]
    public void ReadAsyncThrowsIfCancelled()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            var cancel = new CancellationTokenSource();
            cancel.Cancel();
            var buffer = new byte[1];
            Assert.ThrowsAsync<TaskCanceledException>(async () => await tee.ReadAsync(buffer, 0, 1, cancel.Token));
        }
    }

    [Test]
    public async Task ReadAsyncWaitsForFirstByte()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            var source = new byte[1] { 3 };
            var result = new byte[1];
            var taskRead = tee.ReadAsync(result, 0, 1);
            await tee.WriteAsync(source, 0, 1);
            taskRead.Wait();
            Assert.AreEqual(source, result);
            Assert.AreEqual(1, taskRead.Result);
        }
    }

    [Test]
    public void ReadAsyncDoesNotReturnZeroUntilAtEndIsTriggered()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            var result = new byte[1];
            var taskRead = tee.ReadAsync(result, 0, 1);
            tee._firstByteReceived = true;
            tee._atEnd = true;
            taskRead.Wait();
            Assert.AreEqual(0, taskRead.Result);
        }
    }

    [Test]
    public void CanWriteReturnsTrue()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.AreEqual(true, tee.CanWrite);
        }
    }

    [Test]
    public async Task WriteAsyncWritesToSinks()
    {
        var sample = new byte[10] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        using (var sink1 = new MemoryStream(10))
        using (var sink2 = new MemoryStream(10))
        using (var tee = new TeeStream(sink1, sink2))
        {
            await tee.WriteAsync(sample, 0, 10);

            var result = new byte[10];
            sink1.Position = 0;
            await sink1.ReadAsync(result, 0, 10);
            Assert.AreEqual(sample, result);
            sink2.Position = 0;
            await sink2.ReadAsync(result, 0, 10);
            Assert.AreEqual(sample, result);
        }
    }

    [Test]
    public async Task WriteAsyncSetsFirstByteReceived()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.False(tee._firstByteReceived);
            await tee.WriteAsync(new byte[] { 1 }, 0, 1);
            Assert.True(tee._firstByteReceived);
        }
    }

    [Test]
    public async Task WriteAsyncWaitsForOutputBuffer()
    {
        var sample = new byte[2048];
        var rnd = new Random();
        rnd.NextBytes(sample);

        using (var tee = new TeeStream(1024, TeeStream.Self))
        {
            // Write first block
            await tee.WriteAsync(sample, 0, 1024);

            // Write second block, which will not fit in buffer until we read
            var taskWrite2 = tee.WriteAsync(sample, 1024, 1024);
            var result = new byte[2048];
            await tee.ReadAsync(result, 0, 1024);
            taskWrite2.Wait();

            // Read and validate remainder
            await tee.ReadAsync(result, 1024, 2048);
            Assert.AreEqual(sample, result);
        }
    }

    [Test]
    public async Task WriteAsyncThrowsIfCancelled()
    {
        var cancel = new CancellationTokenSource();
        cancel.Cancel();

        var sample = new byte[2048];
        var rnd = new Random();
        rnd.NextBytes(sample);

        using (var tee = new TeeStream(1024, TeeStream.Self))
        {
            await tee.WriteAsync(sample, 0, 1024);
            Assert.ThrowsAsync<TaskCanceledException>(async () =>
                await tee.WriteAsync(sample, 1024, 1024, cancel.Token));
        }
    }

    [Test]
    public async Task FlushAsyncCallsSinkAsync()
    {
        var sink1 = new Mock<MemoryStream>();
        var sink2 = new Mock<MemoryStream>();
        sink1.CallBase = true;
        sink2.CallBase = true;
        sink1.SetupGet(s => s.CanWrite).Returns(true);
        sink2.SetupGet(s => s.CanWrite).Returns(true);
        sink1.Setup(s => s.Flush()).Verifiable();
        sink2.Setup(s => s.Flush()).Verifiable();
        using (var tee = new TeeStream(sink1.Object, sink2.Object))
        {
            await tee.FlushAsync();
            sink1.Verify(s => s.Flush(), Times.Once);
            sink2.Verify(s => s.Flush(), Times.Once);
        }
    }

    [Test]
    public void PositionThrowsNotImplemented()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            long x;
            Assert.Throws<NotImplementedException>(() => x = tee.Position);
            Assert.Throws<NotImplementedException>(() => tee.Position = 1);
        }
    }

    [Test]
    public async Task ReadCallsReadAsync()
    {
        var tee = new Mock<TeeStream>(TeeStream.Self);
        tee.CallBase = true;
        var buffer = new byte[1] { 1 };
        await tee.Object.WriteAsync(buffer, 0, 1);
        tee.Setup(t => t.ReadAsync(It.Is<byte[]>(b => b.Equals(buffer)),
            It.Is<int>(i => i.Equals(0)),
            It.Is<int>(i => i.Equals(1)),
            It.Is<CancellationToken>(c => c.Equals(CancellationToken.None)))).Verifiable();
        tee.Object.Read(buffer, 0, 1);
        tee.Verify(t => t.ReadAsync(It.Is<byte[]>(b => b.Equals(buffer)),
            It.Is<int>(i => i.Equals(0)),
            It.Is<int>(i => i.Equals(1)),
            It.Is<CancellationToken>(c => c.Equals(CancellationToken.None))), Times.Once);
    }

    [Test]
    public void WriteCallsWriteAsync()
    {
        var tee = new Mock<TeeStream>(TeeStream.Self);
        tee.CallBase = true;
        var buffer = new byte[1] { 1 };
        tee.Setup(t => t.WriteAsync(It.Is<byte[]>(b => b.Equals(buffer)),
            It.Is<int>(i => i.Equals(0)),
            It.Is<int>(i => i.Equals(1)),
            It.Is<CancellationToken>(c => c.Equals(CancellationToken.None)))).Verifiable();
        tee.Object.Write(buffer, 0, 1);
        tee.Verify(t => t.WriteAsync(It.Is<byte[]>(b => b.Equals(buffer)),
            It.Is<int>(i => i.Equals(0)),
            It.Is<int>(i => i.Equals(1)),
            It.Is<CancellationToken>(c => c.Equals(CancellationToken.None))), Times.Once);
    }

    [Test]
    public void FlushCallsFlushAsync()
    {
        var tee = new Mock<TeeStream>(TeeStream.Self);
        tee.CallBase = true;
        tee.Setup(t => t.FlushAsync(It.Is<CancellationToken>(c => c.Equals(CancellationToken.None)))).Verifiable();
        tee.Object.Flush();
        tee.Verify(t => t.FlushAsync(It.Is<CancellationToken>(c => c.Equals(CancellationToken.None))), Times.Once);
    }

    [Test]
    public void LengthThrowsNotImplemented()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            long x;
            Assert.Throws<NotImplementedException>(() => x = tee.Length);
        }
    }

    [Test]
    public void SetLengthThrowsNotImplemented()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.Throws<NotImplementedException>(() => tee.SetLength(0));
        }
    }


    [Test]
    public void CanSeekReturnsFalse()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.False(tee.CanSeek);
        }
    }

    [Test]
    public void SeekThrowsNotImplemented()
    {
        using (var tee = new TeeStream(TeeStream.Self))
        {
            Assert.Throws<NotImplementedException>(() => tee.Seek(0, SeekOrigin.Begin));
        }
    }
}

src/TeeStream.cs
namespace TeeStreaming;

using System;
using System.Buffers;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

/// <summary>
/// A utility to "tee" a source stream to one or more sink streams
/// </summary>
public class TeeStream : Stream
{
    public readonly static Stream Self = Stream.Null;

    protected internal IEnumerable<Stream> _sinks;
    protected internal RingBuffer? _outputBuffer;
    protected internal int _bufferSize;
    protected internal bool _firstByteReceived = false;
    protected internal bool _atEnd = false;

    protected long _totalBytesRead = 0;
    protected long _totalBytesWritten = 0;

    public const int DEFAULT_TEESTREAM_BUFFER_SIZE = 81920;

    /// <summary>
    /// Convenience constructor
    /// </summary>
    /// <param name="sinks"></param>
    public TeeStream(params Stream[] sinks) : this(DEFAULT_TEESTREAM_BUFFER_SIZE, sinks)
    {
    }

    /// <summary>
    /// Instantiate and configure a StreamTee
    /// </summary>
    /// <param name="sinks">Writable stream sinks</param>
    /// <param name="bufferSize">Size of buffer</param>
    /// <exception cref="ArgumentException"></exception>
    public TeeStream(int bufferSize = DEFAULT_TEESTREAM_BUFFER_SIZE, params Stream[] sinks)
    {
        var sinkCount = sinks.Length;
        if (sinkCount == 0)
        {
            throw new ArgumentException("There must be at least one sink");
        }
        if (bufferSize < 1024)
        {
            throw new ArgumentException("Buffer size must be at least 1,024");
        }

        var idx = 1;
        foreach (var sink in sinks)
        {
            if (sink == TeeStream.Self)
            {
                _outputBuffer = new RingBuffer(bufferSize);
            }
            else if (!sink.CanWrite)
            {
                throw new ArgumentException($"Sink #{idx} is not writable");
            }
            idx++;
        }
        _bufferSize = bufferSize;
        _sinks = sinks.Where(sink => sink != TeeStream.Self);
    }

    /// <summary>
    /// Close the Tee stream and any sink streams as well
    /// </summary>
    public override void Close()
    {
        base.Close();
        foreach (var sink in _sinks)
        {
            sink.Close();
        }
    }

    /// <summary>
    /// Tidy up
    /// </summary>
    /// <param name="disposing"></param>
    protected override void Dispose(bool disposing)
    {
        base.Dispose(disposing);
        _outputBuffer?.Dispose();
        _outputBuffer = null;
    }

    /// <summary>
    /// Copy from the source stream into our tee.  Unlike CopyTo, this allows us to know when the end of the stream is.
    /// We also use switch between two buffers to facilitate simultaneous read/write operations
    /// </summary>
    /// <param name="source"></param>
    /// <param name="bufferSize"></param>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public async Task CopyFromAsync(Stream source, CancellationToken? cancellationToken = null)
    {
        await Task.Run(async () =>
        {
            var holdBufferSize = _bufferSize / 2;
            var holdThreshold = holdBufferSize * 90 / 100;
            byte[] holdBufferA = ArrayPool<byte>.Shared.Rent(holdBufferSize);
            byte[] holdBufferB = ArrayPool<byte>.Shared.Rent(holdBufferSize);

            var bufferToRead = holdBufferA;
            var bufferToWrite = holdBufferB;

            var bufferToReadLength = 0;
            var bufferToWriteLength = 0;

            var tasks = new List<Task>(2);

            try
            {
                var token = cancellationToken.HasValue ? cancellationToken.Value : CancellationToken.None;

                int bytesRead;
                do
                {
                    var readTask = source.ReadAsync(bufferToRead, bufferToReadLength, holdBufferSize - bufferToReadLength, token);
                    tasks.Add(readTask);
                    if (bufferToWriteLength > 0)
                    {
                        tasks.Add(WriteAsync(bufferToWrite, 0, bufferToWriteLength, token));
                        bufferToWriteLength = 0;
                    }

                    Task.WaitAll(tasks.ToArray());

                    bytesRead = readTask.Result;
                    
                    bufferToReadLength += bytesRead;

                    if (bufferToReadLength > holdThreshold && bytesRead != 0)
                    {
                        var thisBuffer = (bufferToRead == holdBufferA) ? holdBufferA : holdBufferB;
                        var otherBuffer = (bufferToRead == holdBufferA) ? holdBufferB : holdBufferA;

                        bufferToWriteLength = bufferToReadLength;
                        bufferToWrite = thisBuffer;

                        bufferToReadLength = 0;
                        bufferToRead = otherBuffer;
                    }

                } while (bytesRead > 0);

                if (bufferToReadLength > 0) {
                    await WriteAsync(bufferToRead, 0, bufferToReadLength, token);
                }

                _atEnd = true;
            }
            finally
            {
                ArrayPool<byte>.Shared.Return(holdBufferA);
                ArrayPool<byte>.Shared.Return(holdBufferB);
            }
        }).ConfigureAwait(false);
    }

    /// <summary>
    /// Call this after the final WriteAsync to indicate the end of an input stream
    /// </summary>
    public void SetAtEnd()
    {
        _atEnd = true;
    }

    /// <summary>
    /// Indicates stream can be read from if configured for self streaming
    /// </summary>
    public override bool CanRead => _outputBuffer != null;

    /// <summary>
    /// If configured for self streaming, copies content in the output buffer
    /// </summary>
    /// <param name="buffer"></param>
    /// <param name="offset"></param>
    /// <param name="count"></param>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    /// <exception cref="NotImplementedException"></exception>
    public override Task<int> ReadAsync(Byte[] buffer, int offset, int count, CancellationToken cancellationToken)
    {
        return Task.Run(() =>
        {
            if (_outputBuffer == null)
            {
                throw new NotImplementedException();
            }

            // System.Console.Error.WriteLine($"Read called with {count} bytes");
            // Make sure we we don't return zero because we haven't received the first
            // byte yet, or we have cleared out the buffer but have not yet received
            // the "at end" signal
            while ((!_firstByteReceived) || ((!_atEnd) && (_outputBuffer.Length == 0)))
            {
                if (cancellationToken.IsCancellationRequested)
                {
                    throw new TaskCanceledException();
                }
                Thread.Yield();
            }

            var result = _outputBuffer.Read(buffer, offset, count);
            _totalBytesRead += result;
            // Console.Error.WriteLine($"Bytes Read: {result}, Length: {_outputBuffer.Length}, Total: {_totalBytesRead}");
            return result;
        });
    }

    /// <summary>
    /// Indicates that stream can be written to
    /// </summary>
    public override bool CanWrite { get => true; }

    /// <summary>
    /// Forwards the buffer to the Tee's sinks and, if configured, the Tee' self output
    /// </summary>
    /// <param name="buffer"></param>
    /// <param name="offset"></param>
    /// <param name="count"></param>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
    {
        return Task.Run(() =>
        {
            // System.Console.Error.WriteLine($"Write called for {count} bytes");
            // Write out data to any defined sinks
            Task.WaitAll(_sinks.Select(sink => sink.WriteAsync(buffer, offset, count, cancellationToken)).ToArray());

            // If self-outputing, then update buffer once there is room
            if (_outputBuffer != null)
            {
                var bytesRemaining = count;
                while (bytesRemaining > 0)
                {
                    int bytesToWrite;
                    do
                    {
                        bytesToWrite = Math.Min(_outputBuffer.Available, bytesRemaining);
                        if (bytesToWrite == 0)
                        {
                            if (cancellationToken.IsCancellationRequested)
                            {
                                throw new TaskCanceledException();
                            }
                            Thread.Yield();
                        }
                    } while (bytesToWrite < 1);
                    _outputBuffer.Write(buffer, offset, bytesToWrite);
                    offset += bytesToWrite;
                    _totalBytesWritten += bytesToWrite;
                    // Console.Error.WriteLine($"Bytes Written: {bytesToWrite}, Length: {_outputBuffer.Length}, Total: {_totalBytesWritten}");
                    bytesRemaining -= bytesToWrite;
                    _firstByteReceived = true;
                }
            }
        });
    }

    /// <summary>
    /// Flush any sink streams
    /// </summary>
    /// <param name="cancellationToken"></param>
    /// <returns></returns>
    public override Task FlushAsync(CancellationToken cancellationToken)
    {
        return Task.Run(() =>
        {
            Task.WaitAll(_sinks.Select(sink =>
            {
                var task = sink.FlushAsync(cancellationToken);
                task.ConfigureAwait(false);
                return task;
            }).ToArray());
        });
    }

    /// <summary>
    /// Position is not implemented in TeeStream
    /// </summary>
    public override long Position
    {
        get => throw new NotImplementedException();
        set => throw new NotImplementedException();
    }

    /// <summary>
    /// Perform a blocking read of the stream
    /// </summary>
    /// <param name="buffer"></param>
    /// <param name="offset"></param>
    /// <param name="count"></param>
    /// <returns></returns>
    /// <exception cref="InvalidOperationException"></exception>
    public override int Read(byte[] buffer, int offset, int count)
    {
        var task = Task.Run(() => ReadAsync(buffer, offset, count, CancellationToken.None));
        task.Wait();
        return task.Result;
    }

    /// <summary>
    /// Perform a blocking write of the stream
    /// </summary>
    /// <param name="buffer"></param>
    /// <param name="offset"></param>
    /// <param name="count"></param>
    /// <exception cref="InvalidOperationException"></exception>
    public override void Write(byte[] buffer, int offset, int count)
    {
        Task.Run(() => WriteAsync(buffer, offset, count, CancellationToken.None)).Wait();
    }

    /// <summary>
    /// Perform a blocking flush of the stream
    /// </summary>
    /// <exception cref="InvalidOperationException"></exception>
    public override void Flush()
    {
        Task.Run(() => FlushAsync(CancellationToken.None)).Wait();
    }

    /// <summary>
    /// Length is not implemented in TeeStream
    /// </summary>
    public override long Length
    {
        get => throw new NotImplementedException();
    }

    /// <summary>
    /// SetLength is not implemented in TeeStream
    /// </summary>
    /// <param name="length"></param>
    /// <exception cref="NotImplementedException"></exception>
    public override void SetLength(long length)
    {
        throw new NotImplementedException();
    }

    /// <summary>
    /// Indicates that Seek is not implemented in TeeStream
    /// </summary>
    public override bool CanSeek { get => false; }

    /// <summary>
    /// Seek is not implemented in TeeStream
    /// </summary>
    /// <param name="offset"></param>
    /// <param name="origin"></param>
    /// <returns></returns>
    /// <exception cref="NotImplementedException"></exception>
    public override long Seek(long offset, SeekOrigin origin)
    {
        throw new NotImplementedException();
    }
}